File: packages/kotlin/src/main/kotlin/me/really/crypto/Hmac.kt
// SPDX-FileCopyrightText: Copyright © 2026 ReallyMe LLC. All rights reserved
//
// SPDX-License-Identifier: Apache-2.0

package me.really.crypto

import java.security.GeneralSecurityException
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

public object ReallyMeHmac {
    public const val MAX_KEY_LENGTH: Int = 4096
    public const val SHA256_TAG_LENGTH: Int = 32
    public const val SHA384_TAG_LENGTH: Int = 48
    public const val SHA512_TAG_LENGTH: Int = 64

    public fun authenticateSha256(key: ByteArray, message: ByteArray): ByteArray =
        authenticate(ProviderAlgorithm.HMAC_SHA256, key, message)

    public fun authenticateSha384(key: ByteArray, message: ByteArray): ByteArray =
        authenticate(ProviderAlgorithm.HMAC_SHA384, key, message)

    public fun authenticateSha512(key: ByteArray, message: ByteArray): ByteArray =
        authenticate(ProviderAlgorithm.HMAC_SHA512, key, message)

    public fun verifySha256(tag: ByteArray, key: ByteArray, message: ByteArray): Boolean {
        if (tag.size != SHA256_TAG_LENGTH) {
            throw ReallyMeCryptoException.InvalidInput()
        }
        return verifyAndClearExpectedTag(tag, authenticateSha256(key, message))
    }

    public fun verifySha384(tag: ByteArray, key: ByteArray, message: ByteArray): Boolean {
        if (tag.size != SHA384_TAG_LENGTH) {
            throw ReallyMeCryptoException.InvalidInput()
        }
        return verifyAndClearExpectedTag(tag, authenticateSha384(key, message))
    }

    public fun verifySha512(tag: ByteArray, key: ByteArray, message: ByteArray): Boolean {
        if (tag.size != SHA512_TAG_LENGTH) {
            throw ReallyMeCryptoException.InvalidInput()
        }
        return verifyAndClearExpectedTag(tag, authenticateSha512(key, message))
    }

    private fun authenticate(
        algorithm: ProviderAlgorithm,
        key: ByteArray,
        message: ByteArray,
    ): ByteArray {
        validateKey(key)
        return try {
            val mac = Mac.getInstance(algorithm.jceName)
            mac.init(SecretKeySpec(key, algorithm.jceName))
            mac.doFinal(message)
        } catch (_: GeneralSecurityException) {
            throw ReallyMeCryptoException.ProviderFailure()
        }
    }

    private fun validateKey(key: ByteArray) {
        if (key.isEmpty() || key.size > MAX_KEY_LENGTH) {
            throw ReallyMeCryptoException.InvalidInput()
        }
    }

    private fun constantTimeEquals(left: ByteArray, right: ByteArray): Boolean {
        if (left.size != right.size) {
            return false
        }
        var difference = 0
        for (index in left.indices) {
            difference = difference or (left[index].toInt() xor right[index].toInt())
        }
        return difference == 0
    }

    private fun verifyAndClearExpectedTag(tag: ByteArray, expectedTag: ByteArray): Boolean =
        try {
            constantTimeEquals(tag, expectedTag)
        } finally {
            // JCE returns a new managed array. Clear it immediately so verification
            // does not retain key-derived authentication material until a GC cycle.
            expectedTag.fill(0)
        }

    private enum class ProviderAlgorithm(val jceName: String) {
        HMAC_SHA256("HmacSHA256"),
        HMAC_SHA384("HmacSHA384"),
        HMAC_SHA512("HmacSHA512"),
    }
}

File: packages/ts/src/hmac.ts
// SPDX-FileCopyrightText: Copyright © 2026 ReallyMe LLC. All rights reserved
//
// SPDX-License-Identifier: Apache-2.0

import { hmac } from "@noble/hashes/hmac.js";
import { sha256, sha384, sha512 } from "@noble/hashes/sha2.js";
import { ReallyMeCryptoError } from "./errors.js";

export const HMAC_MAX_KEY_LENGTH = 4096;
export const HMAC_SHA256_TAG_LENGTH = 32;
export const HMAC_SHA384_TAG_LENGTH = 48;
export const HMAC_SHA512_TAG_LENGTH = 64;

export const ReallyMeHmac = {
  authenticateSha256(key: Uint8Array, message: Uint8Array): Uint8Array {
    validateKey(key);
    return hmac(sha256, key, message);
  },

  authenticateSha384(key: Uint8Array, message: Uint8Array): Uint8Array {
    validateKey(key);
    return hmac(sha384, key, message);
  },

  authenticateSha512(key: Uint8Array, message: Uint8Array): Uint8Array {
    validateKey(key);
    return hmac(sha512, key, message);
  },

  verifySha256(tag: Uint8Array, key: Uint8Array, message: Uint8Array): boolean {
    if (tag.length !== HMAC_SHA256_TAG_LENGTH) {
      throw new ReallyMeCryptoError("invalid-input");
    }
    return verifyAndClearExpectedTag(tag, ReallyMeHmac.authenticateSha256(key, message));
  },

  verifySha384(tag: Uint8Array, key: Uint8Array, message: Uint8Array): boolean {
    if (tag.length !== HMAC_SHA384_TAG_LENGTH) {
      throw new ReallyMeCryptoError("invalid-input");
    }
    return verifyAndClearExpectedTag(tag, ReallyMeHmac.authenticateSha384(key, message));
  },

  verifySha512(tag: Uint8Array, key: Uint8Array, message: Uint8Array): boolean {
    if (tag.length !== HMAC_SHA512_TAG_LENGTH) {
      throw new ReallyMeCryptoError("invalid-input");
    }
    return verifyAndClearExpectedTag(tag, ReallyMeHmac.authenticateSha512(key, message));
  },
} as const;

function validateKey(key: Uint8Array): void {
  if (key.length === 0 || key.length > HMAC_MAX_KEY_LENGTH) {
    throw new ReallyMeCryptoError("invalid-input");
  }
}

function constantTimeEquals(left: Uint8Array, right: Uint8Array): boolean {
  if (left.length !== right.length) {
    return false;
  }
  let difference = 0;
  for (const [index, leftByte] of left.entries()) {
    const rightByte = right[index];
    if (rightByte === undefined) {
      return false;
    }
    difference |= leftByte ^ rightByte;
  }
  return difference === 0;
}

function verifyAndClearExpectedTag(tag: Uint8Array, expectedTag: Uint8Array): boolean {
  try {
    return constantTimeEquals(tag, expectedTag);
  } finally {
    // Noble returns a new typed array. Clear it immediately so verification
    // does not retain key-derived authentication material until a GC cycle.
    expectedTag.fill(0);
  }
}
